### ISSUE TITLE
Escapaing dollars requires two back-slashes

### ISSUE THREAD

**Issue by choldgraf** [jupyterlab/jupyterlab#8645]
I am not sure whether this is expected behavior or not, but I find it strange so wanted to bring up here.

It seems that dollar signs are only escaped if they have **two** backslashes. E.g. if you wanted to write a sentence like this:

This costs $200 and that costs $300

Here's how it behaves in jupyterlab and the notebook:

raw:
![image](https://user-images.githubusercontent.com/1839645/86518330-e972d980-bde5-11ea-83e0-a2625857d6ab.png)

rendered:
![image](https://user-images.githubusercontent.com/1839645/86518332-f0015100-bde5-11ea-97f1-df5ac41af052.png)

It feels like a single backslash should be enough to escape dollar signs...any idea what's going on?

**Comment by blink1073** [comment]
Sounds like a math parsing limitation, here's where we're doing it: https://github.com/jupyterlab/jupyterlab/blob/f0153e0258b32674c9aec106383ddf7b618cebab/packages/rendermime/src/latex.ts#L16

**Comment by chrisjsewell** [comment]
+1

**Comment by lesteve** [comment]
As discussed in https://github.com/executablebooks/jupyter-book/issues/750#issuecomment-774141567
1. Jupyter notebook classic interface as the same behaviour so this is not JupyterLab specific
2. pandoc seems to agree that `\$` is the right way to escape a dollar sign and `\\$` is not

For some more context: we bumped into this bug in JupyterBook because there is currently no way to have a syntax for escaping the dollar sign that works both inside a notebook and inside the JupyterBook HTML output.

**Comment by telamonian** [comment]
Tracing the issue further, the initial pass of our math parsers seem to work fine, but then any single backslashes in front of dollar signs will get dropped as a result of the following line:

https://github.com/jupyterlab/jupyterlab/blob/32c98cc091537566e79d98a40c9950381e38978a/packages/rendermime/src/renderers.ts#L343-L344

**Comment by choldgraf** [comment]
I'm curious if @bollwyvl ran into this issue in the jupyterlab-markup work?

**Comment by krassowski** [comment]
Regardless of the rendering issue, highlighting is off too. It can be addressed by requiring that open/close delimiters for tex mode are not directly preceded by (odd/even) number of `\`; the relevant code is here:

https://github.com/jupyterlab/jupyterlab/blob/cec0d84ea5877a8766a2146bac5e0f0ad0d590f1/packages/codemirror/src/codemirror-ipythongfm.ts#L69-L72

**Comment by JasonWeill** [comment]
Still an issue in Lab 4.1.0.

**Comment by dlqqq** [comment]
This issue can be fixed with a minor modification to the implementation of `@jupyterlab/mathjax-extension`, but it's unclear whether this would be a breaking change. 

The MathJax object is initialized with custom configuration in `packages/mathjax-extension/src/index.ts`:

```ts
      ...
      const tex = new TeX({
        packages: AllPackages.concat('require'),
        inlineMath: [
          // ! vvvv ! THIS LINE SPECIFIES $ AS AN INLINE DELIMITER
          ['$', '$'],
          ['\\(', '\\)']
        ],
        displayMath: [
          ['$$', '$$'],
          ['\\[', '\\]']
        ],
        processEscapes: true,
        processEnvironments: true
      });

      const mathDocument = mathjax.document(window.document, {
        InputJax: tex,
        OutputJax: chtml
      });
```

As highlighted above, JupyterLab explicitly lists `$` as a valid delimiter for inline math. However, this differs from MathJax's default configuration, which does not recognize `$` as a delimiter for inline math. This was done intentionally by the authors of MathJax, as the `$` symbol is frequently used to literally denote quantities of USD. The upstream documentation has a dedicated page explaining this design decision: https://docs.mathjax.org/en/latest/input/tex/delimiters.html

In my effort to implement https://github.com/jupyterlab/jupyter-ai/pull/1068, I manually updated the MathJax config to not use `$` as an inline delimiter. The below screenshot shows how math markup would change in JupyterLab if the highlighted line above were removed:

<img width="1037" alt="Screenshot 2024-10-30 at 3 43 26 PM" src="https://github.com/user-attachments/assets/a96db3d8-70e6-48e1-822a-2dd397127d1e">

**Comment by dlqqq** [comment]
The reason double backslashes are necessary is two-fold:

1. `\$` and `\(` are being recognized as [escape sequences](https://en.wikipedia.org/wiki/Escape_sequence) by the JupyterLab rendermime extension, so these strings are displayed literally as `$` and `(` when rendered respectively.

2. MathJax directly reads & writes to the DOM, which now lack the backslashes which were removed by the JupyterLab renderer.

Hence, backslashes need to be doubled, such that `\\(` gets rendered as `\(` by the rendermime extension, which then gets read by MathJax.

**Comment by dlqqq** [comment]
Given all of this context, I've outlined a basic plan for getting this issue fixed.

1. Configure the rendermime extension to not treat `\(`, `\)`, `\[`, `\]`, and `\$` as escape sequences. In other words, these inputs should be rendered literally as-is.
    - If this isn't possible, an alternative strategy can be attempted: double the backslashes immediately before rendering (i.e. changing `\(` to `\\(` in-memory, such that a user input of `\(` is rendered as `\(`). 
2. Remove support for `$` as an inline delimiter, as described in a previous comment.

These code changes would result in the following changes to JupyterLab's behavior:

- Users will no longer be able to use `$` to enclose inline math expressions.
- Users will be able to type `$` literally, without double backslashes.
- Users will be able to enclose inline math expressions in the form `\( <math> \)` literally, without double backslashes.
- Users will be able to enclose display math expressions in the form `\[ <math> \]` literally, without double backslashes.

Again, it remains unclear whether these changes are breaking or non-breaking; I will leave that open for discussion.

Contributors, please share your feedback about this and see if you would be open to helping implement this! It would be great to fix this issue for users who use the `$` symbol frequently, e.g. for financial applications.

**Comment by ellisonbg** [comment]
This is a really good callout and would be a simple way of fixing this. Thanks for doing the research. At the same time, the single `$` delimeter for LaTeX is very common in notebooks (my own brain is wired to it, and many users coming from LaTeX use this). Because of that, this change would likely break many existing notebooks. At a minimum, we would need to do this in a major version bump of the notebook format, JupyterLab, and Jupyter Notebook. Also, non-Jupyter notebook frontends (Colab, VSCode, etc.) would also need to make changes to accommodate this. Because of this, and the millions of existing notebooks, I don't believe this is a viable fix.

**Comment by krassowski** [comment]
> Again, it remains unclear whether these changes are breaking or non-breaking; I will leave that open for discussion.

I agree with Brain, this is very much a breaking change. We never break user notebooks. It _should_ be easy to do in a major version bump of nbformat, but would probably deserve to be a part of a larger solution where markdown format is specified (so that the renderer knows whether to use katex, mathjax 2, mathjax 3, etc and with which options), see https://github.com/jupyter/enhancement-proposals/issues/98.

**Comment by krassowski** [comment]
@dlqqq https://github.com/jupyterlab/jupyterlab/pull/19177 might be of your interest - I would appreciate a look
